Allow tournament client to use default mod icons if custom icons are not present

MiraiSubject opened the pull request:
This implements game side's mod icons into the tournament client. Tournament managers can still choose to use their own assets to override but if they are not found it will fall back to the original mod icon.

As far as I know there's no native NoMod icon, so that will currently stay blank in the Mappool/Round Editor. 

Example of what it will look like in the panel for Hidden: 
![image](https://user-images.githubusercontent.com/803255/105638622-e61f6800-5e73-11eb-9edd-e6b345ab184b.png)

MiraiSubject committed:
Implement native osu!lazer mod icons for tournament

bdach commented on osu.Game.Tournament/Components/TournamentBeatmapPanel.cs:
```diff
@@ -192,5 +185,86 @@ private void updateState()
+
+                    switch (Mod)
+                    {
+                        case "DT":
+                            selectedMod = new OsuModDoubleTime();
+                            break;
+
+                        case "FL":
+                            selectedMod = new OsuModFlashlight();
+                            break;
+
+                        case "HT":
+                            selectedMod = new OsuModHalfTime();
+                            break;
+
+                        case "HD":
+                            selectedMod = new OsuModHidden();
+                            break;
+
+                        case "HR":
+                            selectedMod = new OsuModHardRock();
+                            break;
+
+                        case "NF":
+                            selectedMod = new OsuModNoFail();
+                            break;
+
+                        case "EZ":
+                            selectedMod = new OsuModEasy();
+                            break;
```
should use `Mod.Acronym`

frenzibyte replied:
resolving ruleset bindable and retrieve mod that matches acronym would be even more safe (for ruleset-specific mods to work as well).

MiraiSubject commented:
Will also include the updated project reference in an incoming commit, once the review is addressed. Forgot to include that.

frenzibyte commented on osu.Game.Tournament/Components/TournamentBeatmapPanel.cs:
```diff
@@ -192,5 +185,86 @@ private void updateState()
             }
         }
+
+        private class ModSprite : Container
```
should probably be:
 1. moved to a separate file.
 2. renamed to `TournamentModDisplay` or something of the sorts.
 3. inherit `CompositeDrawable` instead of `Container`, shouldn't be able to mutate children from outside that component.

frenzibyte commented on osu.Game.Tournament/Components/TournamentBeatmapPanel.cs:
```diff
@@ -192,5 +185,86 @@ private void updateState()
+
+            public ModSprite()
+            {
+                Margin = new MarginPadding(10);
+                Width = 60;
+                RelativeSizeAxes = Axes.Y;
```
these are specific for the usage in panel, should be inlined instead of set inside component.

MiraiSubject committed:
Change ModSprite to use ruleset's mods directly.

frenzibyte commented on osu.Game.Tournament/Components/TournamentBeatmapPanel.cs:
```diff
@@ -124,21 +126,11 @@ private void load(LadderInfo ladder, TextureStore textures)
-                        Origin = Anchor.CentreRight,
-                        Texture = textures.Get($"mods/{mods}"),
-                    }
+                    Mod = mods
```
I'm quite confused by this set, is `mods` always meant to be one mod? can't this be renamed to `mod` then?

MiraiSubject replied:
Yeah it was always one mod but it was always called `mods` afaik.  I should rename it, considering it will still be one mod until we can support multiple mods in a panel. 

https://github.com/ppy/osu/blob/f2236312a0323d8a68117c60ae3e46400e75c27e/osu.Game.Tournament/Components/TournamentBeatmapPanel.cs#L140

frenzibyte replied:
Change `ModDisplay` to `ModIcon` while at it, simplifies code.

MiraiSubject commented:
Thanks @frenzibyte for the pointer. I've addressed the review and it's now referencing the acronyms from the ruleset itself.

MiraiSubject committed:
Remove unneccessary ruleset parameter

frenzibyte reviewed (changes requested):
Few more things to note, mostly nits on code quality/readability, and also I would highly recommend adding test cases for these changes if possible, as they're really helpful in the reviewing process, writing one for this should be pretty much simple.

peppy commented:
It should only support a single mod for now, correct.

Also just from the screenshot, the default ones look larger than expected? They should roughly match the sizing of the OWC tournament ones (by force, so they can't exceed the panel size etc.).

MiraiSubject committed:
Split and rename TournamentModDisplay component

MiraiSubject committed:
Change parameter to be singular mod instead of plural

MiraiSubject committed:
Rename mod to modIcon

MiraiSubject committed:
Add visual test for Tournament Mod Display

MiraiSubject committed:
Remove unncessary test step

MiraiSubject committed:
Add null checks to prevent nullrefexception in automated test

MiraiSubject committed:
use null propragtor for Ruleset.Value and rulset instead of null checks

MiraiSubject committed:
Merge branch 'master' into osu-mod-icon-fallbacks

MiraiSubject commented:
fwiw the mod icons used in OWC 2020 differ from the game side mod icons. 

![image](https://user-images.githubusercontent.com/803255/105698769-a190dc80-5f06-11eb-89aa-7027e87eb187.png)

I have however tried to match the scale of the icons somewhat closely 

![image](https://user-images.githubusercontent.com/803255/105715417-deb49900-5f1d-11eb-9969-f3adafc83911.png)

I have also addressed the review and added a visual test (as can be seen in above screenshot).

frenzibyte commented on osu.Game.Tournament/Components/TournamentModDisplay.cs:
```diff
@@ -0,0 +1,57 @@
+{
+    public class TournamentModDisplay : CompositeDrawable
+    {
+        public string ModAcronym;
```
move value set to constructor, this shouldn't be mutable from outside

MiraiSubject committed:
move value set to constructor and make private readonly

peppy commented:
That's looking a lot closer to what I'd expect.

bdach commented on osu.Game.Tournament/Components/TournamentModDisplay.cs:
```diff
@@ -0,0 +1,62 @@
+                if (modIcon == null)
+                    return;
+
+                AddInternal(new ModIcon(modIcon)
+                {
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    Scale = new Vector2(0.5f)
+                });
```
The mod icon has a tooltip, which is possibly unintended in a tournament client context (might end up hovering on it by accident and blocking out other elements of the screen). Should probably be disabled.

peppy committed:
Avoid tooltip display

peppy committed:
Rename class and add xmldoc

peppy committed:
Tidy up nesting
